src/function/execute.rs

```diff
@@ -38,53 +38,87 @@ where
-                            C::recover_from_cycle(db, &cycle, C::id_to_input(db, id))
-                        } else {
-                            // we are not a participant in this cycle
-                            debug_assert!(!cycle
-                                .participant_keys()
-                                .any(|k| k == database_key_index));
-                            cycle.throw()
+        let mut iteration_count = 0;
+
+        loop {
+            let active_query = zalsa_local.push_query(database_key_index);
+
+            // If we already executed this query once, then use the tracked-struct ids from the
+            // previous execution as the starting point for the new one.
+            if let Some(old_memo) = &opt_old_memo {
+                active_query.seed_tracked_struct_ids(&old_memo.revisions.tracked_struct_ids);
+            }
+
+            // Query was not previously executed, or value is potentially
+            // stale, or value is absent. Let's execute!
+            let mut new_value = C::execute(db, C::id_to_input(db, id));
+            let mut revisions = active_query.pop();
+
+            // If the new value is equal to the old one, then it didn't
+            // really change, even if some of its inputs have. So we can
+            // "backdate" its `changed_at` revision to be the same as the
+            // old value.
+            if let Some(old_memo) = &opt_old_memo {
+                self.backdate_if_appropriate(old_memo, &mut revisions, &new_value);
+                self.diff_outputs(db, database_key_index, old_memo, &mut revisions);
+            }
+
+            // Did the new result we got depend on our own provisional value, in a cycle?
+            if revisions.cycle_heads.contains(&database_key_index) {
+                if let Some(last_provisional) = opt_last_provisional {
+                    if let Some(provisional_value) = &last_provisional.value {
+                        // If the new result is equal to the last provisional result, the cycle has
+                        // converged and we are done.
+                        if !C::values_equal(&new_value, provisional_value) {
+                            // We are in a cycle that hasn't converged; ask the user's
+                            // cycle-recovery function what to do:
+                            match C::recover_from_cycle(db, &new_value, iteration_count) {
+                                crate::CycleRecoveryAction::Iterate => {
+                                    iteration_count += 1;
+                                    revisions.cycle_ignore = false;
+                                    opt_last_provisional = Some(self.insert_memo(
+                                        zalsa,
+                                        id,
+                                        Memo::new(Some(new_value), revision_now, revisions),
+                                    ));
+                                    continue;
```

Comment:
Yes, this is one of the areas where we may want to spend some more time on performance optimization. Reusing memos is tricky because in general once they are in the memo map, unless we have a mutable reference to the db (which we don't have here, we only have in between revisions), we have to assume someone else may have a reference to that memo.

Reply:
Yeah, that's true. I wonder if we can make some use of the knowledge that these memo's can only be reached by tracked functions participating in a cycle. They would have to store the memo value somewhere globally for it to escape and that would require casting the value to 'static?

Reply:
Another thing to try here would be to avoid pre-emptively pushing a provisional memo for every query that supports fixpoint, and instead lazily generate and insert that provisional memo only if/when a cycle is encountered. This would really reduce the overhead in the scenario "queries support fixpoint but relatively few cycles actually encountered," which I think is the scenario whose performance we should be most concerned about here.

I think this might be doable; I'll experiment with it once I have more thorough tests in place as guardrails.

Reply:
This is now done; it turned out to be the same thing that I needed in order to fix multiple-revisions.